File: IsometricRPG/IsometricRPG/World/ChunkCoord.swift
/// Identifies a chunk in the infinite grid. Hashable so we can use it as a dictionary key.
struct ChunkCoord: Hashable, Equatable {
    let x: Int
    let y: Int

    /// The world-space origin (top-left tile) of this chunk.
    var worldOriginCol: Int { x * Constants.chunkSize }
    var worldOriginRow: Int { y * Constants.chunkSize }

    /// Deterministic seed for this chunk derived from world seed + position.
    var chunkSeed: UInt64 {
        var seed = Constants.worldSeed
        seed ^= UInt64(bitPattern: Int64(x)) &* 0x517CC1B727220A95
        seed ^= UInt64(bitPattern: Int64(y)) &* 0x6C62272E07BB0142
        seed ^= seed >> 33
        seed &*= 0xFF51AFD7ED558CCD
        seed ^= seed >> 33
        return seed
    }

    /// Manhattan distance to another coord.
    func distance(to other: ChunkCoord) -> Int {
        return abs(x - other.x) + abs(y - other.y)
    }

    /// Chebyshev distance (max of axis distances).
    func chebyshevDistance(to other: ChunkCoord) -> Int {
        return max(abs(x - other.x), abs(y - other.y))
    }

    /// Which chunk contains the given world position.
    static func containing(worldCol: Int, worldRow: Int) -> ChunkCoord {
        // Use floor division to handle negative coordinates
        let cx = worldCol >= 0 ? worldCol / Constants.chunkSize : (worldCol - Constants.chunkSize + 1) / Constants.chunkSize
        let cy = worldRow >= 0 ? worldRow / Constants.chunkSize : (worldRow - Constants.chunkSize + 1) / Constants.chunkSize
        return ChunkCoord(x: cx, y: cy)
    }

    static func containing(worldPosition: CGPoint) -> ChunkCoord {
        return containing(worldCol: Int(floor(worldPosition.x)),
                         worldRow: Int(floor(worldPosition.y)))
    }
}

File: IsometricRPGMono/World/ChunkCoord.cs
namespace IsometricRPG;

public struct ChunkCoord : IEquatable<ChunkCoord>
{
    public int X { get; }
    public int Y { get; }

    public ChunkCoord(int x, int y)
    {
        X = x;
        Y = y;
    }

    /// <summary>
    /// Converts a world-space tile position to its containing chunk coordinate.
    /// </summary>
    public static ChunkCoord FromWorld(int tileX, int tileY)
    {
        return new ChunkCoord(
            (int)Math.Floor((double)tileX / Constants.ChunkSize),
            (int)Math.Floor((double)tileY / Constants.ChunkSize)
        );
    }

    /// <summary>
    /// Returns the world-space tile origin (top-left corner) of this chunk.
    /// </summary>
    public (int tileX, int tileY) WorldOrigin()
    {
        return (X * Constants.ChunkSize, Y * Constants.ChunkSize);
    }

    /// <summary>
    /// Manhattan distance to another chunk.
    /// </summary>
    public int ManhattanDistance(ChunkCoord other)
    {
        return Math.Abs(X - other.X) + Math.Abs(Y - other.Y);
    }

    // IEquatable<ChunkCoord>
    public bool Equals(ChunkCoord other) => X == other.X && Y == other.Y;
    public override bool Equals(object? obj) => obj is ChunkCoord c && Equals(c);
    public override int GetHashCode() => HashCode.Combine(X, Y);

    public static bool operator ==(ChunkCoord a, ChunkCoord b) => a.Equals(b);
    public static bool operator !=(ChunkCoord a, ChunkCoord b) => !a.Equals(b);

    public override string ToString() => $"ChunkCoord({X}, {Y})";
}
